### PR TITLE
feat: instrument mysql2 prepare statement

### PR DESCRIPTION
**Pull request by xuan-cao-swi**
### Description

This PR aims to enhance mysql2 instrumentation by capturing prepare statement execution, akin to the [pg instrumentation](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/opentelemetry-instrumentation-pg/v0.27.1/instrumentation/pg/lib/opentelemetry/instrumentation/pg/patches/connection.rb#L30-L37).

In contrast to the pg instrumentation, this implementation for mysql2 focuses solely on instrumenting prepare statements without extending the instrumentation to [execute statements](https://github.com/brianmario/mysql2/blob/0.5.5/lib/mysql2/statement.rb#L3). The rationale behind this approach is that the mysql2 client passes only the binding values to the execute function, offering minimal informative value, which is why the execute function is not instrumented. Additionally, another reason for this distinction is the way active_record executes these [two types of function executions](https://github.com/rails/rails/blob/v7.1.3/activerecord/lib/active_record/connection_adapters/mysql2/database_statements.rb#L123-L128) separately.

### Reproduce
database.yml
```
default: &default
  adapter: mysql2
  pool: <%= ENV.fetch("RAILS_MAX_THREADS") { 5 } %>
  timeout: 5000

development:
  <<: *default
  host: sample-mysql
  port: 3306
  database: classicmodels
  username: root
  password: admin
  prepared_statements: true
```  
Execute the query:
```ruby
Customer.where("contactLastName like ?", 'Schmitt')
# OR
stmt = "SELECT contactLastName from customers where contactLastName like ?"
binds = [[nil, 'Schmitt']]
result = ActiveRecord::Base.connection.exec_query stmt, 'SQL', binds, prepare: true
```
Sample span
```
#<struct OpenTelemetry::SDK::Trace::SpanData name="select", kind=:client, status=#<OpenTelemetry::Trace::Status:0x0000ffff7f562618 @code=1, @description="">, parent_span_id="A\x92\xE9S\xCC\x0E\xFD\xA8", total_recorded_attributes=5, total_recorded_events=0, total_recorded_links=0, start_timestamp=1707448523002822050, end_timestamp=1707448523003312883, attributes={"db.system"=>"mysql", "net.peer.name"=>"sinatra-mysql", "net.peer.port"=>"3306", "db.name"=>"classicmodels", "db.statement"=>"SELECT `customers`.* FROM `customers` WHERE (contactLastName like ?) ORDER BY `customers`.`customerNumber` ASC LIMIT ?"}, links=nil, events=nil, resource=#<OpenTelemetry::SDK::Resources::Resource:0x0000ffff7f206dc0 @attributes={"service.name"=>"unknown_service", "process.pid"=>805, "process.command"=>"/usr/local/bundle/bin/rackup", "process.runtime.name"=>"ruby", "process.runtime.version"=>"3.1.0", "process.runtime.description"=>"ruby 3.1.0p0 (2021-12-25 revision fb4df44d16) [aarch64-linux]", "telemetry.sdk.name"=>"opentelemetry", "telemetry.sdk.language"=>"ruby", "telemetry.sdk.version"=>"1.4.0"}>, instrumentation_scope=#<struct OpenTelemetry::SDK::InstrumentationScope name="OpenTelemetry::Instrumentation::Mysql2", version="0.26.1">, span_id="\xAF\xDC\f\xC0\xEA_\xD9W", trace_id="\x92]'\x18\xDB4J5[\xCC\a\xE2\xA7\x8E\xD0\xAF", trace_flags=#<OpenTelemetry::Trace::TraceFlags:0x0000ffff7f441e78 @flags=1>, tracestate=#<OpenTelemetry::Trace::Tracestate:0x0000ffff7f28ebf8 @hash={}>>
```

Related to issue https://github.com/open-telemetry/opentelemetry-ruby-contrib/issues/670

### CODE REVIEW

File: instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/patches/client.rb [hunk 1]
```diff
@@ -15,6 +15,45 @@ module Patches
...
+              attributes: attributes.merge!(OpenTelemetry::Instrumentation::Mysql2.attributes),
+              kind: :client
+            ) do
+              super(sql)
+            end
+          end
+
+          private
+
+          def span_name(sql)
```
Comment (arielvalentin) [inline instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/patches/client.rb:44]:
```suggestion
          def _otel_span_name(sql)
```
Something that we haven't done yet and probably should is adopt a naming convention for private methods we add via monkey patches to a class.

We should start prefixing all private methods with `_otel` going forward.

File: instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/patches/client.rb [hunk 2]
```diff
@@ -15,6 +15,45 @@ module Patches
         module Client
           def query(sql, options = {})
             attributes = client_attributes
+            attributes.merge!(span_attributes(sql))
+
+            tracer.in_span(
+              span_name(sql),
```
Comment (arielvalentin) [inline instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/patches/client.rb:21]:
```suggestion
              _otel_span_name(sql),
```

File: instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/patches/client.rb [hunk 3]
```diff
@@ -15,6 +15,45 @@ module Patches
...
+              super(sql, options)
+            end
+          end
+
+          def prepare(sql)
+            attributes = client_attributes
+            attributes.merge!(span_attributes(sql))
+
+            tracer.in_span(
+              span_name(sql),
```
Comment (arielvalentin) [inline instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/patches/client.rb:34]:
```suggestion
              _otel_span_name(sql),
```

File: instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/patches/client.rb [hunk 4]
```diff
@@ -15,6 +15,45 @@ module Patches
...
+              sql,
+              OpenTelemetry::Instrumentation::Mysql2.attributes[
+                SemanticConventions::Trace::DB_OPERATION
+              ],
+              database_name,
+              config
+            )
+          end
+
+          def span_attributes(sql)
```
Comment (arielvalentin) [inline instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/patches/client.rb:55]: If you're able to update this as well please

**Review by arielvalentin — APPROVED** [review]
Small suggestion that doesn't have to be a part of this PR but I think we should adopt going forward.

File: instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/patches/client.rb (hunk 4 above)
Comment (xuan-cao-swi) [reply instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/patches/client.rb]: Thanks for the review, I have changed all the private method with `_otel`, but let me if some of the private method is not required with `_otel`.

**Review by arielvalentin — APPROVED** [review]
Thank you for this contribution!

**Pull request merged**